File: bst/BST/BST.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace BST
{
    internal class BST
    {
        Node<int> Head;

        public void AddNode(int nodeValue)
        {
            if (Head == null)
            {
                Node<int> newNode = new Node<int>(nodeValue);
                Head = newNode;
                return;
            }
            AddNode(nodeValue, Head);
        }
        public void AddNode(int nodeValue, Node<int> currentNode = null)
        {
            Node<int> newNode = new Node<int>(nodeValue);

            if (newNode.Value < currentNode.Value)
            {
                if (currentNode.Left == null)
                {
                    currentNode.Left = newNode;
                    return;
                }
                AddNode(nodeValue, currentNode.Left);
            }

            if (newNode.Value > currentNode.Value)
            {
                if (currentNode.Right == null)
                {
                    currentNode.Right = newNode;
                    return;
                }
                AddNode(nodeValue, currentNode.Right);
            }
        }

        public int DeleteNode(int nodeValue, Node<int> currentNode = null)
        {
            if (currentNode == null)
            {
                currentNode = Head;
            }

            if (nodeValue < currentNode.Value)
            {
                if (currentNode.Left.Value == nodeValue)
                {
                    Node<int> nodeToDelete = currentNode.Left;

                    if (nodeToDelete.Left == null && nodeToDelete.Right == null)
                    {
                        currentNode.Left = null;
                    }
                    if (nodeToDelete.Left == null || nodeToDelete.Right == null)
                    {
                        var childNode = nodeToDelete.Left;
                        if (childNode == null) { childNode = nodeToDelete.Right; }

                        nodeToDelete.Right = null;
                        nodeToDelete.Left = null;

                        currentNode.Left = childNode;
                    }
                }
                return DeleteNode(nodeValue, currentNode.Left);
            }

            if (nodeValue > currentNode.Value)
            {
                if (currentNode.Right.Value == nodeValue) { }
                return DeleteNode(nodeValue, currentNode.Right);
            }
            return currentNode.Value;
        }

        //private Node<int> FindParent(int targetValue, Node<int> currentNode)
        //{

        //}
    }

    class Node<T>
    {
        public T Value;
        public Node<T> Left;
        public Node<T> Right;

        public Node(T Value) { }
    }
}

File: bst/bst.php
<?php

class Node 
{
    public int $value;
    public ?Node $left = null;
    public ?Node $right= null;

    public function __construct(int $value)
    {
        $this->value = $value;
    }
}

class BST 
{
    public ?Node $head = null;

    public function addNode(int $value, ?Node $currentNode = null): void
    {
        if ($this->head == null)
        {
            $this->head = new Node($value);
            return;
        } 

        if( $currentNode == null){
            $currentNode = $this->head;
        } 

        if ( $value > $currentNode->value )
        {
            if( $currentNode->right == null) $currentNode->right = new Node($value);
            
            else $this->addNode($value, $currentNode->right);
        } 
        elseif ( $value < $currentNode->value)
        {
            if( $currentNode->left == null ) $currentNode->left = new Node($value);

            else $this->addNode($value, $currentNode->left);
        }
    }

    public function removeNode(int $val, ?Node $currentNode = null): void
    {
        if($currentNode == null) $currentNode = $this->head;

        if ( $currentNode->right->value == $val ){
            $nodeLeft = $currentNode->right->left;
            $nodeRight = $currentNode->right->right;
        
            if( $nodeLeft == null )
            {
                $currentNode->right = $nodeRight;
            }
            elseif ( $nodeRight == null)
            {
                $currentNode->right = $nodeLeft;
            }
            else 
            {

            }
        }
        elseif( $currentNode->left->value == $val ){
            $nodeLeft = $currentNode->left->left;
            $nodeRight = $currentNode->left->right;
        
            if( $nodeLeft == null )
            {

            }
            elseif ( $nodeRight == null)
            {

            }
            else 
            {

            }
        }

        if( $val > $currentNode->value )
        {
            $this->removeNode($val, $currentNode->right);
        }
        elseif ($val < $currentNode->value)
        {
            $this->removeNode($val, $currentNode->left);
        } 
        else 
        {
            $nodeLeft = $currentNode->left;
            $nodeRight = $currentNode->right;
        
            if( $nodeLeft == null )
            {

            }
            elseif ( $nodeRight == null)
            {

            }
            else 
            {

            }
        }
    }

    public function inOrder(?Node $currentNode = null): void
    {
        if($currentNode == null)
        {
            $currentNode = $this->head;
        }
        if($currentNode->left)
        {
            $this->inOrder($currentNode->left);
        }
        echo $currentNode->value . "\n";

        if($currentNode->right) $this->inOrder($currentNode->right);
    }
}


$bst = new BST();

$bst->addNode(4);
$bst->addNode(8);
$bst->addNode(3);
$bst->addNode(10);
$bst->addNode(11);
$bst->addNode(1);

// echo $bst->head->value;
// echo $bst->head->right->value;

$bst->inOrder();